tools/serverpod_cli/bin/generator/types.dart

```diff
@@ -0,0 +1,348 @@
+                            'Map.fromEntries((data as List).map((e) =>'
+                            'MapEntry(deserializeJson<'),
+                        generics.first.reference(serverCode).code,
+                        const Code('>(e[\'k\']),deserializeJson<'),
+                        generics[1].reference(serverCode).code,
+                        const Code('>(e[\'v\']))))' ':null) as dynamic')
+                      ])
+                    : Block.of([
+                        // using Code.scope only sets the generic to List
+                        const Code('Map.fromEntries((data as List).map((e) =>'
+                            'MapEntry(deserializeJson<'),
+                        generics.first.reference(serverCode).code,
+                        const Code('>(e[\'k\']),deserializeJson<'),
+                        generics[1].reference(serverCode).code,
+                        const Code('>(e[\'v\'])))) as dynamic')
+                      ])
+          ]),
+        ),
+        ...generics.first.generateDeserialization(serverCode),
+        ...generics[1].generateDeserialization(serverCode),
+      ];
+    } else if (customClass) {
+      return [
+        MapEntry(
+            nullable
+                ? refer('getType', serverPodUrl(serverCode))
+                    .call([], {}, [reference(serverCode)])
+                : reference(serverCode),
+            Code.scope((a) => nullable
+                ? '(data!=null?'
+                    '${a(reference(serverCode))}.fromJson(data,this)'
+                    ':null)as T'
+                : '${a(reference(serverCode))}.fromJson(data,this) as T'))
+      ];
+    } else {
+      return [];
+    }
+  }
+}
+
+/// Analyze the type at the start of [input].
+/// [input] must not contain spaces.
+/// Returns a [_TypeResult] containing the type,
+/// as well as the position of the last parsed character.
+/// So when calling with "List<List<String?>?>,database",
+/// the position will point at the ','.
+/// If [analyzingCustomConstructors] is true, the root element might be marked as [TypeDefinition.customClass].
+TypeParseResult parseAndAnalyzeType(String input,
+    {bool analyzingCustomConstructors = false}) {
+  String classname = '';
+  for (var i = 0; i < input.length; i++) {
+    switch (input[i]) {
+      case '<':
+        var generics = <TypeDefinition>[];
+        while (true) {
+          i++;
+          var result = parseAndAnalyzeType(input.substring(i));
+          generics.add(result.type);
+          i += result.parsedPosition;
+          if (input[i] == '>') {
+            var nullable = (i + 1 < input.length) && input[i + 1] == '?';
+            return TypeParseResult(
+                i + 1,
+                TypeDefinition.mixedUrlAndClassName(
+                  mixed: classname,
+                  nullable: nullable,
+                  generics: generics,
+                ));
+          } else if (i >= input.length - 1) {
+            throw 'INVALID TYPE';
```

Comment:
Please use `SourceSpanException` (ideally) or `FormatException`. (Generally, never throw anything other than Exceptions.)

Reply:
Like this? 3071a5e

Reply:
Yes, that looks correct! The SouceSpanException can provide better feedback to the user. :)